src/scanner.c

```diff
@@ -2895,29 +2893,24 @@ void Pr (
   PrTo(GetCurrentOutput(), format, arg1, arg2);
 }
 
-/* TL: static Char *theBuffer; */
-/* TL: static UInt theCount; */
-/* TL: static UInt theLimit; */
+typedef struct {
+    Char * TheBuffer;
+    UInt   TheCount;
+    UInt   TheLimit;
+} BufferState;
 
-static void putToTheBuffer( Char c)
+static void putToTheBuffer(void *state, Char c)
 {
```

Comment:
I find the naming of the parameter `state` a bit confusing, I scrolled through the whole code to even find out what it does. Could we not call the struct `PrintBuffer` and the parameter `buffer`? Why is it a `void *` anyway?

Reply:
The argument is a `void` so that this function has the signature which `FormatOutput` needs -- we can of course change it here, but then we need to typecast the function pointer. 

The argument is named `state` because that's the generic term used in the argument to `FormatOutput` and the other function we pass to `FormatOutput`, but I am happy to change it to `buffer` here if you prefer. I guess a better "generic" name than `state` might be `stream`, as that is essentially the role of this argument (indeed, if you then also swap the two arguments, you essentially arrive at the signature of `fputc`, which calls its arguments `c` and `stream`.

Not sure why you "scrolled through the whole code to find what it does" -- this function is only referenced once, 10 lines after it is being declared; and it immediately casts its argument.

Note: I am not saying this code is "clear", I just don't understand what exactly the problem is?!

Reply:
But, you decided FormatOutput should take a void*.. why not just have everything to PrintBuffer, if that is all these functions will ever take / work on? Or are you preparing for future possible expansion?

Reply:
But that is *not* all they ever take. There are two calls, with different putchar pointers. Only one writes to a buffer.